### ISSUE TITLE
Option to export vertex colors as non-color data

### ISSUE THREAD

**Issue by jitspoe** [KhronosGroup/glTF-Blender-IO#542]
I'm currently using vertex colors as data, so I want to preserve the raw values as I enter them in blender.  For example, I might use one color channel to determine how much sway foliage has for a vertex shader.  However, if I enter 0.5, for 50% sway, after the sRGB->linear conversion happens, I end up with a value of ~0.22, meaning I have to compensate for this when I color the verts, or I have to do a linear->srgb conversion in the shader, and I'd prefer not to add more calculations.  Also, there's data loss if I convert the values when I try to set them in Blender, since it appears that the colors are stored as 8bits per channel.

Note: Original issue filed was incorrect (I had colorspaces backward in my head), and this became repurposed, so I edited this.  Leaving the original post below:

The glTF exporter runs vertex colors through a function called color_srgb_to_scene_linear().  There are a couple issues with this:
1) I'm using vertex colors as data, not colors, so I want to preserve the raw values.
2) color_srgb_to_scene_linear() seems to be backward:
```
def color_srgb_to_scene_linear(c):
    if c < 0.04045:
        return 0.0 if c < 0.0 else c * (1.0 / 12.92)
    else:
        return pow((c + 0.055) * (1.0 / 1.055), 2.4)
```
srgb->linear conversion should be using 1/2.4.  Basically seems like these function names are reversed.

**To Reproduce**
1. Export using the glTF format.
2. Note the vertex color values are LOWER than the values in blender

**Expected behavior**
If values stored in the glTF format are supposed to be linear, and Blender is using sRGB, the sRGB -> linear conversion should make the values HIGHER/BRIGHTER.  If Blender is already using linear values (not sure what the Blender vertex colorspace is), no conversion should take place. Would be nice to have a checkbox for export saying "I'm already using Linear vertex color values" to avoid conversion.

**Version**
 - Blender Version [e.g. 2.80 (2019-05-17)]

**Additional context**
Looks like this is where the color conversion was introduced: https://github.com/KhronosGroup/glTF-Blender-IO/issues/261

**Comment by donmccurdy** [comment]
Let's start with the question about whether `color_srgb_to_scene_linear` is correct. Vertex colors in glTF are supposed to be linear, and I'm fairly confident Blender's Python API is returning sRGB values (even though it's _hopefully_ using linear values internally). I haven't found that documented anywhere, but testing seems to confirm below.

I'm not sure why you think the conversion is backward, or what you mean when you say it should make the values brighter – you should see no difference when viewing colors in any correct glTF implementation. An easy way to check is to create a *Background* material (unaffected by lighting), export it, and check the result in a couple of viewers:

Using a cube with a solid `#4285F4` vertex coloring, and disabling the default *Filmic* view transform:

![Screen Shot 2019-06-14 at 10 53 10 AM](https://user-images.githubusercontent.com/1848368/59528302-c00da580-8e92-11e9-9876-7067d753dd78.png)

![Screen Shot 2019-06-14 at 10 53 44 AM](https://user-images.githubusercontent.com/1848368/59528299-bd12b500-8e92-11e9-97b8-a5e8bf7e3e10.png)
![Screen Shot 2019-06-14 at 10 53 46 AM](https://user-images.githubusercontent.com/1848368/59528304-c1d76900-8e92-11e9-93ab-9eeb1b16bbc4.png)

So I think the the implementation is correct – if you're getting different results, then it would help us if you can provide a `.blend` file, or let us know what viewer you're opening the file in – perhaps that renderer is doing the wrong thing?

If we have correctness sorted out, we can consider options for vertex data. I'm not sure yet how to handle that.

**Comment by jitspoe** [comment]
I apologize.  For some reason I had myself fully convinced that when you go from sRGB->linear that things get brighter (using a < 1.0 exponent value).  Lots of misleading images out there.

![linear_vs_srgbgamma-1000x711](https://user-images.githubusercontent.com/7751182/59548910-630e0000-8f23-11e9-92df-1225d36b3cdf.jpg)
![Linear Lighting-0](https://user-images.githubusercontent.com/7751182/59548911-64d7c380-8f23-11e9-897e-75d34c6b07be.jpg)
![Linear_Distribution_versus_Gamma_Corrected_Distribution](https://user-images.githubusercontent.com/7751182/59548930-95b7f880-8f23-11e9-9ee3-f96bf703255d.jpg)

Of course, a linear display is brighter.  Also, if you do a gradient in linear space, then convert it back to sRGB, it's brighter than if you do a gradient in sRGB gamma space.

Of course, if you're converting an image to display on sRGB, you need to make it brighter to display on the darker screen.  I guess I was thinking about the curve sRGB displays apply to an image.  Anyway, I had it all backward.

The thing that's problematic about the way things currently work is that I'm using the vertex colors as data.  For example, I might use a vertex color to specify how much a tree leaf should sway in a vertex shader, so I want to set a color to 0.5.  I did some tests.  I set the RGB channels to .25, .5, and .75, and made the cube display as flat white:

![image](https://user-images.githubusercontent.com/7751182/59549353-31e4fe00-8f2a-11e9-9cde-5d41dc167021.png)

Then I took a screenshot and used the eyedropper.  Looks like those values are sRGB values, since the values I entered for the color are the same as what is displayed (slightly off because of the 8 bits per pixel display depth -- rounded to the nearest 1/255th).

Finally, I edited the python export script to display the values before and after the conversion:

sRGB  : 0.250980 0.501961 0.749020
Linear: 0.051269 0.215861 0.520996

Sadly, this looks like Blender is internally storing the values as 8 bits per channel sRGB.  .25 * 255 = 63.75, rounded to 6 / 255 = 0.2509803921568627.

So this bug is wrong.  The conversion is correct.  The problem is that I can't enter the values in linear space.  My 0.5 turns to ~0.2, which is less than half of the value it should be for the purposes of my shader.  I either have to compensate for this when I enter values or add more calculations to the shader to convert the linear value back to sRGB to get the values I actually entered.  Or hack the exporter to just not apply the conversion.  Maybe it would be good to have an option to not convert the vertex colors to linear space, though you'd have to know not to apply them on an import if you for some reason reimported the asset back into blender.

Argh, this is a mess.  The other problem with this conversion is that, since the values seem to be stored in 8 bit values, doing the conversion when entering values means you lose precision in the lower values, and that's what I use a lot.  Also, half of my assets were exported as .dae files which did not do this conversion, so if I have a bunch of inconsistencies right now.  For me, personally, it would be better to be able to export without conversion.

**Comment by lexaknyazev** [comment]
glTF supports custom per-vertex data as custom attributes. Blender supports vertex data painting, so there should be a way to pass such values without treating them as colors. I'm not sure whether the exporter supports this workflow, though.

**Comment by emackey** [comment]
So, this issue could be considered a feature request for an option to treat vertex colors as non-color data?

**Comment by emackey** [comment]
And if so, would it still export to `COLOR_0` or should it export to a custom attribute?

**Comment by donmccurdy** [comment]
Yeah, we should think of this as a feature request for non-color data. An option labeled "my colors are already linear" would be confusing, since Blender doesn't use them that way. My thought on how to support this would be to:

(1) Require users to rename the vertex attribute to something other than Blender's default (`Col`).
(2) Store the renamed attribute with that custom name – not `COLOR_0` or `COLOR_1` – and without the sRGB->Linear conversion.
(3) Import the same way, if a custom attribute is present.

Optionally, this could be behind an option – not sure if Blender users would ever rename their vertex colors in normal use?

We could do the same with the UV attribute, since that can also be used for custom vec2 data.

![Screen Shot 2019-06-15 at 1 43 55 PM](https://user-images.githubusercontent.com/1848368/59556290-cc6a2f00-8f74-11e9-86cb-6e408c94dd19.png)

The big caveat on this suggestion is... where are you going to load the model, @jitspoe? I know three.js preserves custom attributes in a glTF model, but I have no idea how widely they're supported beyond that. If your import tool requires the data to be in an actual vertex color attribute, this is a bit trickier.

**Comment by jitspoe** [comment]
I'm importing the gltf files into the Godot engine.  I'm not sure if it supports custom vertex data or not.  There is custom data that can be passed to the shader (typically for particles), but I don't know if the importer reads it.

Honestly, I'm fine if you just close this bug, since it wasn't really a bug, and there are a lot more critical things to deal with, like an option to not export control bones and the fact that multiple meshes on the same armature don't work properly.  I still have to export my character with 2.79's better collada exporter.  Seems trying to allow for colors as data might be a bit of a hack, and I can just use a modified export script to get around it.  Perhaps a new issue for this can be created if others have the same workflow issue.

**Comment by donmccurdy** [comment]
As a (non-urgent) feature request, I do like this idea. I've wanted a way to set custom attributes before myself. I'm not really sure how the end-to-end workflow will be, and I'm definitely not familiar with Godot, but we might as well keep this open.

I'm curious if the Blender Python API has other ways to set custom vertex data, like [BMesh CustomData](https://docs.blender.org/api/blender_python_api_current/bmesh.html), that could be used by other addons? If so, that might be a more direct solution (even if it requires other addons to create the data...).

**Comment by rabanti-github** [comment]
Hello,
I have added an option to export vertex colors as sRGB or linear in a fork.
This option would be available as dropdown with the entries "sRGB" (default) and "Linear" under the checkbox "Vertex Colors" (tab "Mesh").
A caveat is that linear may still have rounding errors, since the stored value are floats.

May I submit a pull request?
The code can be seen at: https://github.com/rabanti-github/glTF-Blender-IO/tree/master/addons/io_scene_gltf2 (automatic generated VSCode stuff can be omitted)

**Comment by donmccurdy** [comment]
@rabanti-github Unfortunately I don't think we could accept that as a pull request. The glTF format is specific that vertex colors must be linear, and displaying an option enabling users to create incorrect glTF files will cause issues and confusion.

A solution that I (at least personally) would support is described in https://github.com/KhronosGroup/glTF-Blender-IO/issues/542#issuecomment-502398739.

**Comment by aten-rgb** [comment]
@donmccurdy Is there any plans going forward to implement this? I am having similar problems described in this thread...

https://blenderartists.org/t/how-to-pick-exact-colour-from-a-vertex/1191859/8

For games engine work this seems like a must have. I know people at my studio are keen on learning blender - myself included. For me I'm trying to use them in a shader, something very common in UE4 and want to replicate this in Blender, a guy in the thread suggested a drop down in the attribute/colour node that lets you select which colour space you want the vertex colours to use which sounds like a great and simple to implement (?) idea to me.

**Comment by donmccurdy** [comment]
We (i.e. the authors of this addon) cannot add a dropdown to Blender attribute nodes. You or other users could certainly suggest that on the Blender discussions, of course.

I'd be happy for someone to implement https://github.com/KhronosGroup/glTF-Blender-IO/issues/542#issuecomment-502398739, such that any vertex colors with custom names get exported without sRGB->linear changes, as custom non-color attributes. But unfortunately I don't know when/if I'll have time to implement this myself.

Short of that, if you're able to share one or two `.blend` examples that would use this feature, and are able to test glTF models created from them, that might help someone to make progress here. Thanks!

**Comment by aten-rgb** [comment]
Hi thanks for the reply, I'm sorry I thought this was a place to ask
requests for blender! My bad. I've only been on it a few days and been
scrambling round all over the place trying to get info. This was the
nearest to what I wanted :)

I'll look for a more suitable place to put the request no worries!

**Comment by donmccurdy** [comment]
Proposed fix in https://github.com/KhronosGroup/glTF-Blender-IO/pull/807.

@aten-rgb @rabanti-github @jitspoe any chance you'd be able to test whether that workflow is a solution for you? You'd need to download this ZIP and install it, replacing the default one in Blender:

https://github.com/donmccurdy/glTF-Blender-IO/archive/feat-custom-attributes.zip

**Comment by jitspoe** [comment]
From my understanding, this will set the color values on a different attribute than the "color".  This won't help my use case, as I'm actually using the "color" attribute in a shader.  Using a different attribute would require:
- Giving vertex colors a consistent custom name
- Checking the custom attributes option on export
- Modifying the game engine's importer to read the custom attribute
- Convert that custom attribute to a color

I can almost guarantee I'll forget to do one of those or typo the name and my shader behavior will be slightly off.  It's also really kind of an obscure workflow that requires arcane knowledge to get right.  Also, if I'm going to go through the effort of making the importer read custom attributes, I might as well just make the importer convert linear back to srgb and be done with it. :)

**Comment by donmccurdy** [comment]
To be up front here — we will not be adding an option to skip sRGB->linear conversion when writing Blender vertex colors to a glTF vertex color attribute. I'm willing to help provide another workflow, but it won't be that specific workflow.

Is it not possible to read from a custom attribute in your shader? I'm fairly confident the Godot developers would be happy to support custom attributes, but they probably haven't done so because there is no good workflow for creating them. Once the one-time implementation work is done, your workflow would be:

- Give vertex colors intended to be used as custom attributes a custom name.
- Enable "custom attributes" option on export.
- Use custom attribute in shader, in Godot.

The first two steps are not obvious, I agree, and would have to be discovered in documentation or exporter settings. But then, custom attributes are a not a particularly commonly-used feature.

I'm assuming you do need all three channels of the vertex colors? For users who didn't, another option would be to support exporting Blender weight paint channels to custom attributes.

**Comment by aten-rgb** [comment]
Just FYI, I managed a workaround in the end by copying the green channel
which I wanted in linear space to the alpha channel which seems to be read
from in linear space and using that.

**Comment by kroko** [comment]
May I ask what is the final verdict on this?, by looking around in multiple threads it is unclear.

I'm "colouring" vertices in Blender in separate (other than default _Col_) vertex colour set and the extra vertex attribute buffer is then used in Three.js for some per vertex manipulations. If I set "colour" to be (0.25, 0.50, 0.75) in Blender for the vertex I expect to receive the exact same values in the other end, without the lossy gamma correction.

I'm still using @donmccurdy proposed https://github.com/KhronosGroup/glTF-Blender-IO/issues/542#issuecomment-559918111 as a way to export custom vertex attributes **linearly**.  forked and branched https://github.com/donmccurdy/glTF-Blender-IO/tree/feat-custom-attributes achieves this - it works.  
Pull https://github.com/KhronosGroup/glTF-Blender-IO/pull/807 was closed.  
Any plans on having this func in glTF-Blender-IO master?

**Comment by donmccurdy** [comment]
In my opinion, the suggestion in https://github.com/KhronosGroup/glTF-Blender-IO/issues/1232#issuecomment-704311864 is the most appropriate solution here. It would enable exporting scalar vertex weights as custom, linear vertex attributes. If anyone would like to take that branch, test it further, and make a PR, I think we could merge it.

The changes in https://github.com/KhronosGroup/glTF-Blender-IO/pull/807 are somewhat different, repurposing Blender's vertex colors for something else. That feels like a "2nd choice" option to me, but it does allow vec3 vertex attributes, which is something... do you (or anyone else) need that, or would scalar vertex weights be enough?

**Comment by dennemark** [comment]
Hi,
just yesterday I asked a question in the babylon forum concerning handling of vertex color [Link](https://forum.babylonjs.com/t/gltf-vertex-colors/18622/6) and it seems to closely related to the issues described here. I only have one vertex color set named "Col". It seems Blender is not exporting the colors as floating values and therefore draco compression seems to mess up vertex colors. BabylonJS won't show a correct result, however GLTF Viewer seems to be able to do it. Reading through this thread I am still a bit confused on how I could set the vertex color to non-color floating data.
 
![Comparsion](https://aws1.discourse-cdn.com/business7/uploads/babylonjs/optimized/2X/e/e0c3ad6b658e54056e3515ea2852ae853dd07c44_2_942x750.jpeg)

**Comment by kroko** [comment]
> It seems Blender is not exporting the colors as floating values

Yep, Blender 2.91.2 vanilla exporter exports it in `Uint16Array` (@donmccurdy fork however in `Float32Array`) ant that's it. Noted that while hacking around.

You can always do

```javascript
const bufferFromGltfUint16 = Uint16Array.from(colAtributeObjRef.array);

const bufferConvertedGltfFloat3d = new Float32Array(bufferFromGltfUint16.length);
for (let vIdx = 0, vLen = bufferFromGltfUint16.length; vIdx < vLen; ++vIdx) {
  bufferConvertedGltfFloat3d[vIdx] = bufferFromGltfUint16[vIdx] / (0xFFFF);
}
```

and pass `bufferConvertedGltfFloat3d` to shaders.

**Comment by dennemark** [comment]
Thanks a lot! Did Blender export with Float32Array before or was it always Uint16Array? To my understanding it would be useful to use Float32Array for GLTF export in blender, correct?
I will try out donmccurdy's fork and your script!

**Comment by kroko** [comment]
> In my opinion, the suggestion in [#1232 (comment)](https://github.com/KhronosGroup/glTF-Blender-IO/issues/1232#issuecomment-704311864) is the most appropriate solution here. It would enable exporting scalar vertex weights as custom, linear vertex attributes. If anyone would like to take that branch, test it further, and make a PR, I think we could merge it.

Downloaded https://github.com/scurest/glTF-Blender-IO/tree/custom-data and it works as advertised.  
Due to the fact it is based on vertex paint it is indeed scalar.  

* Using vertex color to pass vec4 linear per vertex (@donmccurdy fork)
    * Having stripes (vec4 packed per vertex) is kind of "what I'm used to" (`glVertexAttribPointer()` from raw OGL land).
    * Meanwhile, having vec4 is suboptimal, both when less than 4 custom values and more than 4 custom values per vertex needed.
    * Semantics are bit off, as sure, it is not about color (at least in my use case).

* Vertex groups + vertex weight to pass float linear per vertex  (@scurest fork)
   * One can have as many vertex groups as needed, each having it's own weight paint, giving whatever amount of custom vertex attributes, one just have to parse them differently (and it can be repacked in vecX stripes when loading in glTF).
   * I mean, having custom attributes kind of dictates that it would mean having also custom shaders, when it comes to Three.js `RawShaderMaterial` is expected, so manual data prep kind of is given anyways.
  * Sure, xyzw attributes are _popular_ and vertex weight allows only visualisation of one dimension in Blender. However with Blender scripting, anything can be done (i.e., merge 3 different vertex weights into vertex color for visual inspection within Blender, etc).

> The changes in #807 are somewhat different, repurposing Blender's vertex colors for something else. That feels like a "2nd choice" option to me, but it does allow vec3 vertex attributes, which is something... do you (or anyone else) need that, or would scalar vertex weights be enough?

That is hacky indeed (and your gives vec4 😉 ). Meanwhile nevertheless your fork gives linear OOB in Three.js data structures while Blender 2.91.2 vanilla exporter gives corrected. I expect vertex colors to be linear (`[0.5, 0.5, 0.5]` in Blender should give `[0.5, 0.5, 0.5]` in Three.js data structures after model is loaded).

> Textures containing color information (.map, .emissiveMap, and .specularMap) always use sRGB colorspace in glTF, while vertex colors and material properties (.color, .emissive, .specular) use linear colorspace. [source](https://threejs.org/docs/#examples/en/loaders/GLTFLoader)

**Comment by kroko** [comment]
> Thanks a lot! Did Blender export with Float32Array before or was it always Uint16Array?

Not sure. Blender 2.91.2 vanilla exporter is setting `componentType` to `5123` which is `UNSIGNED_SHORT` (`FLOAT` is `5126`) in the exported glTF for vertex colors.

> To my understanding it would be useful to use Float32Array for GLTF export in blender, correct?

Indeed, one should use floating point arithmetic when it comes to your GPU 😄

**Comment by scurest** [comment]
Differences between using vertex painting and weight painting

* Vertex painting is per-corner-of-poly (that is, you can paint a different color in every poly that meets a single vertex). Weight painting is per vertex. 
* Blender stores vertex colors as 8-bit internally, which means vertex color precision is limited to about 0.004.

> Did Blender export with Float32Array before or was it always Uint16Array?

It used to be float. It was changed by #1169.

**Comment by donmccurdy** [comment]
> > To my understanding it would be useful to use Float32Array for GLTF export in blender, correct?
> 
> Indeed, one should use floating point arithmetic when it comes to your GPU

GPU APIs allow normalized integer vertex attributes (https://gamedev.stackexchange.com/questions/9792/opengl-vertex-attributes-normalisation), enabling integer values to be mapped to floats automatically, and saving file size and GPU memory. Because Blender itself does not use 32 bits for vertex colors, we saw no reason to store 2x the necessary data in the file. For the viewer/engine loading the file, 16-bit integers work as expected for vertex colors.

> I expect vertex colors to be linear ([0.5, 0.5, 0.5] in Blender should give [0.5, 0.5, 0.5] in Three.js data structures after model is loaded).

Vertex colors in Blender are not interpreted as linear; they use an sRGB transfer function. We have to account for that during export or your visual results will not match in three.js or other viewers.

> [Weight painting] semantics are still bit off, as sure, it is not about weight.

I think this is well within Blender's intentions for weight painting. From the [Weight Painting docs](https://docs.blender.org/manual/en/latest/sculpt_paint/weight_paint/introduction.html#introduction), "It is primarily used for rigging meshes, where the vertex groups are used to define the relative bone influences on the mesh. But we use it also for controlling particle emission, hair density, many modifiers, shape keys, etc."

***

I would recommend a PR based on https://github.com/KhronosGroup/glTF-Blender-IO/issues/1232#issuecomment-704311864 if anyone is interested in doing that. It avoids all of the confusion around using vertex colors for something else.

**Comment by kroko** [comment]
I also vote for https://github.com/KhronosGroup/glTF-Blender-IO/issues/1232#issuecomment-704311864  
Sadly though there is no Weight paint note in shader editor to quickly pack and visualise multiple 0-1 weights as RGB combination, Vertex color had one.

I'm bit ignorant about Blender when it comes to weight painting and vertex group intricacies, but my notes are that the export and reimport is lossy, when it comes to vertex groups per se, but that is to be expected.
* For @scurest fork it should be extra noted that vertex groups have to be assumed as "custom attribute layers for all vertices" and that vertices assigned in Blender to those groups do not matter. Meaning, glTF will have all vertices of the object in the exported "vertex groups a.k.a custom attribute buffer", not only the assigned vertices to that vertex group. Which is desired and and logical as attribute should be equal the total vertices count for simplicity (not treated as lookup map).
* Current @scurest fork works one way. That is, reimporting glTF into Blender does not create vertex groups and weights for vertices within those groups. If it did it would again show the lossy nature of this approach - even if importing glTF, that has some attributes that are type of SCALAR would trigger creation of vertex groups and weight paint values within those vertex groups in Blender, the information about the "original subset of vertices for that vertex group" would be lost, the importing will assign all vertices to each vertex group. Again, logical, as attributes are for all vertexes.
* One just one has to understand that roundtrip loses information, one does not export vertex groups (say array of references to indices), but all vertices. Which will bite if for some reason in future glTF will have "true vertex groups" functionality.

**Comment by julienduroure** [comment]
Hello,
Not sure this feature is still required, as Blender has now Attributes, that is the way to go for exporting data that are not colors.

Can I close this ticket?

**Comment by jitspoe** [comment]
Well, I've just been hacking the exporter every blender update, which is kind of a pain.  Would be nice if there was just an option to export the color values as-is without conversion to linear like I originally requested.  Attributes and other proposed solutions aren't really viable as that data does not get imported into the engine, so I'd have to modify the importer as well as re-author hundreds of assets to use the new format.  Just not worth the effort.  If no color-based solution is going to come from this, then I guess it can be closed and I'll just suck it up for this project.

**Comment by julienduroure** [comment]
> Would be nice if there was just an option to export the color values as-is without conversion to linear like I originally requested

We can't add this option, as the exported files will not fit the glTF 2.0 specification regarding color management.

Another solution is to add an hook. You will be able to manage your conversion on your own in an extension, without the need to hack the original source code.

**Comment by donmccurdy** [comment]
Independent of this issue, we should try to encourage engines to support custom vertex attributes. I don't know of authoring tools other than Blender (and perhaps Houdini?) that have the ability to export that type of custom data without hacks like encoding floats as vertex colors, and it will be better for everyone if we can make that a widely supported thing.

But, I do think trying to create and maintain special workflows around non-color data pretending to be Linear-sRGB components would create pain for us (and other downstream projects in the glTF ecosystem) with time. 

I'm happy to share a batch offline process for either (a) converting GLB vertex color space, or (b) renaming custom attributes as `COLOR_#` if that's helpful to anyone, but I don't feel great about plugging it into a Blender release.

**Comment by scurest** [comment]
@jitspoe 

>  Would be nice if there was just an option to export the color values as-is without conversion to linear like I originally requested.

When you create a color attribute, make it's Data Type "Color" instead of "Byte Color" (this is the default since 3.2). These colors are already stored in linear inside Blender so they get exported as-is without sRGB->Linear conversion.

![](https://user-images.githubusercontent.com/11024420/208782131-f4a865b0-3a06-4fe0-8d17-4282bc621870.png)

**Comment by julienduroure** [comment]
We can now close this ticket, as
- data type 'color' are linear in Blender
- For people that really want to change attribute & vertex color data, even if this will create not valid glTF files, they can use the following hook:

> gather_attribute_change

An example of user : https://github.com/julienduroure/gltf-hooks-examples/blob/main/gltf_attribute_change.py

**Comment by TheSinkingMan** [comment]
I don't know about glTF but when exporting FBX files no combination of domain or data type will lead to accurately exported color data. Could this glTF export funcionality be applied to FBX as well so that we can get non-sRGB corrected values?

**Comment by julienduroure** [comment]
Hello @TheSinkingMan 
This repository is dedicated to glTF import and export.
For any FBX bug report, you should use https://projects.blender.org/blender/blender
For any FBX feature request, you should use https://blender.community/c/rightclickselect/ or https://devtalk.blender.org/
